compiler/semtypes.nim

```diff
@@ -457,7 +457,10 @@ proc semCaseBranch(c: PContext, t, branch: PNode, branchIndex: int,
       # constant sets and arrays are allowed:
       var r = semConstExpr(c, b)
       # for ``{}`` we want to trigger the type mismatch in ``fitNode``:
-      if r.kind notin {nkCurly, nkBracket} or len(r) == 0:
+      if r.kind == nkCurly and len(r) == 0:
```

Comment:
I added empty arrays and extended the test to cover some more cases including the arrays.